final cm code fixes for #5004

JohnCremona opened the pull request:
Two tiny changes, now that (on the beta database) the 'cm' column of ec_nfcurves now always contains 0 or the actual negative discriminant, as the 'cm_status' column holds the sign of the old 'cm' column (+1 for actual CM and -1 for potential).  Specifically, 
   1. in processing the search for specific discriminants we no longer need to append the negatives of the ones requested;
   2. in preparing the web page data we no longer have to replace positive 'cm' value by  their negatives
It will be safe to push this to prod *after* ec_nfcurves has been copied over, and at that opint#5004 can be closed.

roed314 commented:
This looks like it may be a real test failure: `/EllipticCurve/?cm_disc=-11&include_cm=PCMnoCM` should include `14641.1-a1`....

JohnCremona commented:
> This looks like it may be a real test failure: `/EllipticCurve/?cm_disc=-11&include_cm=PCMnoCM` should include `14641.1-a1`....

Thanks for spotting this, I will look into it.

JohnCremona commented:
> /EllipticCurve/?cm_disc=-11&include_cm=PCMnoCM

14641.1-a1 over which field?  I do see curves with that label over 2.0.19.1, but the sort order hides them.  If you add conductor norm = 14641 to the search criteria you see that there are 12 matches.

AndrewVSutherland commented:
I wouldn't worry about errors on devmirror (I haven't copied ec_nfcurves over yet).  But I do think there is a problem with https://beta.lmfdb.xyz/EllipticCurve/?start=50&hst=List&cm_disc=-11&include_cm=PCMnoCM which is including curves with CM (not just PCM), and this is not fixed by this PR, I see the same thing on http://127.0.0.1:37777/EllipticCurve/?cm_disc=-11&include_cm=PCMnoCM

If I look at https://beta.lmfdb.xyz/EllipticCurve/data/2.0.11.1-24025.7-CMa1 (which has CM not just PCM) I see that `cm_type` is set to 1 which I think is correct based on what you wrote in https://github.com/LMFDB/lmfdb/issues/5004#issue-1127010400 (I just updated the schema doc to reflect this).  I don't see a column named `cm_status`, did you mean `cm_type` in what you wrote above?

So I think the search is not filtering on `cm_type` as it should, and indeed if you look at https://github.com/LMFDB/lmfdb/blob/903b104b3b33877b194c29d1231e42d376e0bcc1/lmfdb/ecnf/main.py#L616 you will see that the logic there needs to change.

AndrewVSutherland commented:
@JohnCremona I'm going to go ahead and merge this and fix the problem I noted above in a separate PR, I really want to get back to a state where tests are passing (I'm busy breaking sato_groups and its dangerous to have things in such a confused state).

JohnCremona commented:
Fair enough -- I'll work out what is wrong (and yes I did mean 'cm_type').

AndrewVSutherland commented:
I'm not convinced anything is actually wrong, but let's see if the PR I just submitted now passes the test that was failing (possibly the failure was simply that by including extra results it was pushing the result that the test was looking for onto the next page.

JohnCremona commented:
I have fixed it and will make a new PR.  You are right that I had not revised the 'include_cm' part of the logic, which is now a lot simpler than it was (thought there were several unnecesarily repeated lines already, in fact).   I'll make the PR as soon as I have checked locally that the relevant test pass.  I also updated the schema description for 'cm' which will need review.

JohnCremona commented:
@AndrewVSutherland See my comment on your PR!  I did not see that you had already done this, with almost the identical changes.
